Version 4.0.0 (Breaking)

marcus-j-davies opened the pull request:
**WORK-IN-PROGRESS**


  - Versions
    - ZWave JS Driver Version: 15.3.0
    - ZWave JS Server Version: 3.1.0 (Schema Version 43)

  - Breaking Changes
    - You now need to build your own **PSI** for the platforms you wish to target.
       Having to maintain the build actions/platforms to keep these updated is eating into the time I have left.
       Building a **PSI** is relatively uneventful - please see the READ ME on how to do this.
    - **DownloadPSI** method on the Helper class has been removed.
    - Removed support for **NET45**  and **NET5.0**
      The supported frameworks are as follows:
           - **NET 48**
           - **NET 6.0**
            - **NET 7.0**
            - **NET 80**
            - **NETSTANDARD 2.0**
            - **NETSTANDARD 2.1**
    - **enableSoftReset** has been moved to **features.softReset** on the ZWaveOptions class
    - The **NodeStatistics** arg on the ZWaveNode class event **StatisticsUpdated**  has been renamed to **NodeStatisticsUpdatedArgs**
    - The **ControllerStatistics** arg on the Controller class event **StatisticsUpdated**  has been renamed to **ControllerStatisticsUpdatedArgs**
    - The **InclusionResult** argument has been renamed to **InclusionResultArgs**
    - The **ValueUpdated** event now uses a dedicated class for the args parameter
    - The **ValueNotification** event now uses a dedicated class for the args parameter
    - The **NodeRemoved** event now contains a reason Enum as to why it was removed.
    - The **NetworkHealDone** and **NetworkHealProgress** events have been renamed to  
               **RebuildRoutesDone** and **RebuildRoutesProgress** and not have dedicated classes for the **args**
    - The **BeginExclusion** method now requires an Exclusion Options instance
    - The Node **BeginFirmwareUpdate** method has been renamed to **UpdateFrimware**, and requires a class instance.
    - The Node **FirmwareUpdateProgress** event now passes an args parameter
    - The Node **FirmwareUpdateFinished** event now passes an args parameter
    - Setting the Node **name**, **location** and **keepAwake** values is now only possible with methods for each.
      This is to address some unintentional communication between the Driver runtime and the lib.  
    - The method **GetAllEndpoints** has been removed, and is replaced with an **endpoints** property

  - New Features  
    - Multiple **Driver** instances are now supported 
    - Added ability to save ZWaveOptions to a JSON string and back
    - Added  **Long Range** (LR) support
    - Added **deviceConfig** property to the Controller class.
    - Added **SetRawConfigParameterValue** method to the ZWaveNode class.
    - Added **RefreshValues** method to the ZWaveNode class.
    - Added **RefreshCCValues** method to the ZWaveNode class.
    - Added **WaitForWakeUp** method to the ZWaveNode class.
    - Added **Ping**, method to the ZwaveNode class.
    - Added **StartListeningLogs**, **StopListeningLogs** methods and the associated events to the Driver class.
    - Added **ValueAdded**, **ValueRemoved** events. **ValueRemoved** was never added until now, **ValueAdded**, previously used the **ValueUpdated** event
    - Added **metadata updated** event to the ZWaveNode class.
    - Added **ManuallyIdleNotificationValue** method to the ZWaveNode class.
    - Added **endpointLabel** to the Endpoint class
    - Added **Interview** method, to the ZWaveNode class - this should only be used if  "disableOnNodeAdded" is set to true
    - Allow specifying Refresh Info options, when re-interviewing a node.
    - Added **FirmwareUpdateOTW** method (and supporting events) to update the Controller Firmware
    - Added **SetRFRegion**, **GetRFRegion**  methods to the Controller class
    - Added **SetPowerlevel**, **GetPowerlevel**  methods to the Controller class
    - Added **GetAvailableFirmwareUpdates** methods to the Controller class
    - Added **ConfigManager**  methods to the Driver class
    - Added **FirmwareUpdateOTA** method to the Controller class, to update a node with the fetched Updates via **GetAvailableFirmwareUpdates**
    - Added **HardReset** method to the Driver class  
            **Warning!!!** This will Reset your controller, and will result in a clean network with no included nodes.
    - Added **SoftReset** method to the Driver class  
    - Added the **ccSpecific** property to the **ValueMetadata** class
    - Added the ability to add new server methods to the library during runtime.  
      This is helpful if you want to use a method that is not yet implemented, or to support an older version of an exetrnal server

  - Internal changes
    - Switched to an alternative websocket client package (which is now embedded)
    - Massive structural / performance improvements
    - All event handlers are now executed with their own `Task` instance
    - Better (hopefully) recovery/connection error handling
    - Updated some of the defaults in the Zwave Options to mirror the Driver defaults
    - Internal logic to ensure the server satifies the specified expected schema requested by the library
    
    Shoutout to...
    @spudwebb for the massive amounts of contribution for this release
    and @georgeinva2004 for testing the stuff I haven't  (which is a lot 😬)
    and @AlCalzone for..... Ummm Zwave JS
    and @raman325 for the foundations needed to expose the Driver to other languages

marcus-j-davies committed:
Update Deps, switch to NPM

marcus-j-davies committed:
Node WaitForWakeUp basics (needs checking)

spudwebb committed:
Add possibility to listen to logging events (#30)

marcus-j-davies committed:
Remove NTK, Change Statistics Name, Change Targets, Swap WS Client, Update FW Method

marcus-j-davies committed:
Value Events

marcus-j-davies committed:
Node Removed update

marcus-j-davies committed:
New Demo, Args, Driver Instance

marcus-j-davies committed:
Update CHANGELOG.md

marcus-j-davies committed:
Reduce Reconnect, Demo App Dev

marcus-j-davies committed:
Added Keep alive to Driver, More Demo dev

marcus-j-davies committed:
Demo App  - Heal WIP

marcus-j-davies committed:
Network Heal

marcus-j-davies committed:
Demo : Heal

marcus-j-davies committed:
NET7.0

marcus-j-davies committed:
Update CHANGELOG.md

marcus-j-davies committed:
Runtime: Additional performance gain

marcus-j-davies committed:
Exclusion Enum and Node Interview

marcus-j-davies committed:
Update CHANGELOG.md

marcus-j-davies committed:
Fix exclusion

marcus-j-davies committed:
Demo Main Form design

marcus-j-davies committed:
Firmware, Region, Power Level

marcus-j-davies committed:
Update README.md

marcus-j-davies committed:
Update README.md

marcus-j-davies committed:
Add ping and fix Power Level processing

marcus-j-davies committed:
Update CHANGELOG.md

marcus-j-davies committed:
Update CHANGELOG.md

marcus-j-davies committed:
Schema version check

marcus-j-davies committed:
Merge branch '4.0.0' of https://github.com/zwave-js/ZWaveJS.NET into 4.0.0

marcus-j-davies committed:
OTA  + FWUS

marcus-j-davies committed:
Update CHANGELOG.md

marcus-j-davies committed:
Update CHANGELOG.md

marcus-j-davies committed:
Hard Reset

marcus-j-davies committed:
Merge branch '4.0.0' of https://github.com/zwave-js/ZWaveJS.NET into 4.0.0

marcus-j-davies committed:
Update CHANGELOG.md

marcus-j-davies committed:
Update CHANGELOG.md

marcus-j-davies committed:
Update CHANGELOG.md

marcus-j-davies committed:
Update CHANGELOG.md

marcus-j-davies committed:
Small relocate

marcus-j-davies commented:
@ georgeinva2004

You may want keep close to this PR - I am extremely close to pushing V4 and given your inputs in the past, I'm giving you a head ups.

If you feel up to it, I am happy to build a PSI for you and build a dev lib.

georgeinva2004 commented:
I have been watching. It's very interesting. I was planning on testing the change. An updated PSI would be very helpful.

Thanks for your hard work.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
From: Marcus Davies ***@***.***>
Sent: Wednesday, August 23, 2023 3:15:03 PM
To: zwave-js/ZWaveJS.NET ***@***.***>
Cc: John M. George ***@***.***>; Mention ***@***.***>
Subject: Re: [zwave-js/ZWaveJS.NET] Version 4.0.0 (Breaking) (PR #31)


@georgeinva2004<https://github.com/georgeinva2004>

You may want keep close to this PR - I am extremely close to pushing V4 and given your inputs in the past, I'm giving you a head ups.

If you feel up to it, I am happy to build a PSI for you and build a dev lib.

—
Reply to this email directly, view it on GitHub<https://github.com/zwave-js/ZWaveJS.NET/pull/31#issuecomment-1690505629>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABJAZ4JG5DPEMVTT3VNBDA3XWZJDPANCNFSM6AAAAAA3XEYMA4>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

marcus-j-davies committed:
Attempt to handle unexpected disconnects

marcus-j-davies committed:
Upload test PIS

marcus-j-davies commented:
Of course!

I have uploaded the latest PSI to the root of the repo.
if you do find the time - please do report back on issues 👍

marcus-j-davies committed:
Seems this needs more work

marcus-j-davies committed:
move start_listening_logs + recovery

marcus-j-davies committed:
bump Server

marcus-j-davies committed:
Update Driver.cs

marcus-j-davies committed:
Recovery Improvement

marcus-j-davies committed:
Update CHANGELOG.md

marcus-j-davies committed:
Update CHANGELOG.md

marcus-j-davies committed:
Recovery & UI

marcus-j-davies committed:
Merge branch '4.0.0' of https://github.com/zwave-js/ZWaveJS.NET into 4.0.0

marcus-j-davies committed:
Add ccSpecific to Metadata class

marcus-j-davies committed:
Update Structures.cs

marcus-j-davies committed:
Update CHANGELOG.md

spudwebb commented:
@marcus-j-davies , in ValueMetadata, could we also add properties for unit and steps
https://zwave-js.github.io/node-zwave-js/#/api/valueid?id=number

also I think that for optional number values like min, max, steps, default, those properties type should be int? instead of int

marcus-j-davies commented:
Hi @spudwebb - Agreed!

I have a mental note to look at the Metadata class, as I'm note sure the `states` property is fully correct also.

Will try and get around to it this weekend.

A PR against the v4 branch will be welcome 😉 it will then be merged with the WIP PR for v4

spudwebb commented:
PR created https://github.com/zwave-js/ZWaveJS.NET/pull/35

The states property works well as is for me.

spudwebb committed:
Add unit and steps properties to ValueMetadata class. Make the default, min and max properties nullable. (#35)

marcus-j-davies committed:
Add connect timeout

marcus-j-davies committed:
Oops!

marcus-j-davies committed:
3rd time lucky

marcus-j-davies committed:
Implement better error handling

spudwebb committed:
Add Refresh Values command (#37)

marcus-j-davies committed:
Update CHANGELOG.md

spudwebb committed:
Add SetRawConfigParameterValue method to the ZWaveNode class (#38)

* Add SetRawConfigParameterValue method to the ZWaveNode class

* Create separation between Server and Driver

---------

Co-authored-by: Marcus Davies <55892693+marcus-j-davies@users.noreply.github.com>

marcus-j-davies committed:
Update Driver.cs

@spudwebb

marcus-j-davies committed:
Small thing

marcus-j-davies committed:
Sentry override

marcus-j-davies committed:
Add Method Factory (V4) (#41)

* Create MethodFactory.cs

* Update CHANGELOG.md

marcus-j-davies committed:
Update build work flow

marcus-j-davies committed:
PSI file names in helper

marcus-j-davies committed:
Update BuildPSI.yml

Fix workflow

spudwebb committed:
Make ZWaveNode.endpoints property public and remove GetAllEndpoints method (#42)

* Make ZWaveNode.endpoints property public and remove GetAllEndpoints method

* Fix sets

* Update ZWaveNode.cs

* Update CHANGELOG.md

---------

Co-authored-by: Marcus Davies <55892693+marcus-j-davies@users.noreply.github.com>

marcus-j-davies committed:
Update schema + fixes (v4) (#47)

* Heal Rename, Firmware Fixes, Schema Bump, Bool Converter Removal, JSON Processing

* Update CHANGELOG.md

* move triggers to a task, add deviceConfig to controller

* Heal renames

* Update embedded runtime build routines

* Delete PSI/package-lock.json

* Update package.json

* Update package.json

* Update build

* Update package.json

* Update package.json

* Firmware Update Service + small corrections

* Firmware Improvements

* so so close

AlCalzone committed:
rebuild server.psi

georgeinva2004 commented:
That's great!!!  Thank-you

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
From: Marcus Davies ***@***.***>
Sent: Friday, November 3, 2023 8:30:10 AM
To: zwave-js/ZWaveJS.NET ***@***.***>
Cc: John M. George ***@***.***>; Mention ***@***.***>
Subject: Re: [zwave-js/ZWaveJS.NET] Version 4.0.0 (Breaking) (PR #31)


@georgeinva2004<https://github.com/georgeinva2004>

The latest code is now available, which address's the issue with forcing a license key for non-commercial use.
There is a recent PSI that will need to be used at the root of V4 (Win x64), as I have bumped the Schema

—
Reply to this email directly, view it on GitHub<https://github.com/zwave-js/ZWaveJS.NET/pull/31#issuecomment-1792353824>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABJAZ4KE5QC4ZTM722QVSEDYCTPVFAVCNFSM6AAAAAA3XEYMA6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTOOJSGM2TGOBSGQ>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

marcus-j-davies committed:
Change Message

marcus-j-davies committed:
Read me

marcus-j-davies committed:
Update README.md

marcus-j-davies committed:
Small fixes

spudwebb committed:
Several changes (#54)

* Added Driver.SoftReset method.
Fixed BeginRebuildingRoutes,  StopRebuildingRoutes and RebuildNodeRoutes methods to use the correct commands.
Fixed typo in RebuildRoutesDone event.

* Removed the old command enums for healing the network and nodes.

* Add SoftReset to CHANGELOG

---------

Co-authored-by: Marcus Davies <55892693+marcus-j-davies@users.noreply.github.com>

spudwebb committed:
Added RefreshCCValues method to the ZWaveNode class. (#55)

* Added RefreshCCValues method to the ZWaveNode class.

* Update ZWaveNode.cs

Just adding my note to state I have checked the severe side signature

---------

Co-authored-by: Marcus Davies <55892693+marcus-j-davies@users.noreply.github.com>

marcus-j-davies committed:
Update CHANGELOG.md

spudwebb committed:
Add missing properties to ValueMetadata (#57)

* Add missing properties to ValueMetadata

* Add ValueFormat parameter to ZWJSS_SetRawConfigParameterValue method

* Align SetRawConfigParameterValue with Options object Variants

---------

Co-authored-by: Marcus Davies <55892693+marcus-j-davies@users.noreply.github.com>

spudwebb committed:
Update ZWaveNode.UpdateFirmware method so that its returned result contains the NodeFirmwareUpdateResultArgs (#59)

spudwebb committed:
Refactor library to support multiple drivers. (#62)

* Refactor library to support multiple drivers.

* Patch server class to handle multiple hosts + check for used ports

* Bump schema + swtich to yao-pkg/pkg

---------

Co-authored-by: Marcus Davies <55892693+marcus-j-davies@users.noreply.github.com>

spudwebb committed:
Add Node Found event to the Controller class (#64)

marcus-j-davies committed:
Update README.md

marcus-j-davies committed:
Delete .github/workflows/BuildPSI.yml

marcus-j-davies committed:
Update README.md

marcus-j-davies committed:
Bump schema and embedded socket client (#66)

* Socket & LR

* Fix Versions

* Nullable Fix

* Fix & Test

* Show Embedded runtime Driver console output during debug

spudwebb committed:
Add NodeAlive event to the ZWaveNode class, and add lastSeen property to the NodeStatistics class. (#67)

* Add NodeAlive event to the ZWaveNode class, and add lastSeen property to the NodeStatistics class.

* Add lastSeen to the ZWaveNode and Statistics classes as DateTime?

spudwebb committed:
Possibility to create smart start provision entries from DSK. (#70)

spudwebb committed:
Add ManuallyIdleNotificationValue method to the ZWaveNode class (#73)

spudwebb committed:
Add MetadataUpdated event to the ZWaveNode class. (#75)

marcus-j-davies committed:
Fix up Smart Start (#71)

* Smart Start

* improve build

* add lint scripts

* Bump schema 40

marcus-j-davies committed:
Update ZWaveNode.cs

marcus-j-davies committed:
Delete .github/workflows/release.yml

marcus-j-davies commented:
@spudwebb 

I have bumped the schema - and Im thinking about doing a release on this (v4)
I have combined some of you recent commits.

 - https://github.com/zwave-js/ZWaveJS.NET/pull/31/commits/6b99e30b46a2c01b396374a56c113c27bb058c88
 - https://github.com/zwave-js/ZWaveJS.NET/pull/31/commits/d27570773827eda2d8f4921b7b6dadb5de2f16c7
 - https://github.com/zwave-js/ZWaveJS.NET/pull/31/commits/233343aa6137646ce888f074a422ce4771698082
 
 I merged them into a secondary PR I was working on (to address smart start inconstancies), and merged that PR to V4
 
 This PR - should now be fully updated with your commits, and has the following versions.
 
  - ZWave JS Driver Version: 14.3.7 (only really applies to Embedded setups)
  - ZWave JS Server Version: 1.40.2 (Schema Version 40)
  
if connecting to a lesser version of the schema (since its now based on 40) - just use the override 

```
public Driver(Uri Server, int SchemaVersion = 0, int ServerErrorThrottleTime = 10000)
```

Do you want to cast your eyes over it?

marcus-j-davies commented:
@georgeinva2004 

This MASSIVE update may get released this week - do you want to test it?

remember to build your PSI

```
cd ./PSI
npm install && npm run build
```

rename ./dist/server.exe to ./dist/server.psi, and distrubute the image with the library.

Note: it doesn't yet have the ability to load a PSI from another location.

marcus-j-davies committed:
Fix duplicated signatures

spudwebb commented:
> @spudwebb
> 
> I have bumped the schema - and Im thinking about doing a release on this (v4) I have combined some of you recent commits.
> 
> * [6b99e30](https://github.com/zwave-js/ZWaveJS.NET/commit/6b99e30b46a2c01b396374a56c113c27bb058c88)
> * [d275707](https://github.com/zwave-js/ZWaveJS.NET/commit/d27570773827eda2d8f4921b7b6dadb5de2f16c7)
> * [233343a](https://github.com/zwave-js/ZWaveJS.NET/commit/233343aa6137646ce888f074a422ce4771698082)
> 
> I merged them into a secondary PR I was working on (to address smart start inconstancies), and merged that PR to V4
> 
> This PR - should now be fully updated with your commits, and has the following versions.
> 
> * ZWave JS Driver Version: 14.3.7 (only really applies to Embedded setups)
> * ZWave JS Server Version: 1.40.2 (Schema Version 40)
> 
> if connecting to a lesser version of the schema (since its now based on 40) - just use the override
> 
> ```
> public Driver(Uri Server, int SchemaVersion = 0, int ServerErrorThrottleTime = 10000)
> ```
> 
> Do you want to cast your eyes over it?

It looks good, but I won't be able to test it before next week.

marcus-j-davies committed:
Add ConfigManager to Driver

marcus-j-davies committed:
Validate ConfigManager for correctness

marcus-j-davies committed:
Patch Soft Reset

marcus-j-davies committed:
Patch Inclusion Started event

marcus-j-davies committed:
Add ability to Serialize/Deserialize ZWaveOptions

AlCalzone commented:
@marcus-j-davies you should probably bump the driver to 14.3.8. 14.3.7 has an issue with reading from cache.

marcus-j-davies committed:
Bump driver

marcus-j-davies commented:
Thanks @AlCalzone for the heads up - Bumped!

marcus-j-davies committed:
Update ZWaveOptions.cs

spudwebb committed:
Fix compatibility with zwave-js-server version 3.0.0 and above. Schema version compatibility is now only checked on server side. (#84)

Remove some compilation warnings.

marcus-j-davies committed:
bump to  v15

marcus-j-davies committed:
Fix config

marcus-j-davies committed:
Implement FS Driver, to better handle PKG paths

marcus-j-davies commented:
@spudwebb  - Hope your well?

I'm going to be finishing up V4 today, and going to be moving to v5 right after (Branch already based on v4).
V4, will exist only for backwards compatibility for environments  < .NET 5 (I.e to support your need)

But After V4 - it will drop support for older targets, V5 will get continued updates but will only support the modern targets

 - net6.0
 - net7.0
 - net8.0
 - net9.0
 - netstandard2.1
 - .... Next
 
 I will be removing the embedded sockets library in v5  - as I need to bring it up to the latest code bases.
  If you still need on going support for the latest driver version in V4, I think you may need to start maintaining your fork of V4.
  
I will bring V4 upto the latest Schema of course  - but after that, V5 will be the active branch, where it will steer away from legacy frameworks - hence my suggestion to keep your own fork for V4.

**tl;dr**
V4 will be the last version to support `net48`, `netstandard2.0 `

spudwebb commented:
Fair enough, I will maintain my own fork of v4.

marcus-j-davies commented:
Let me first bring v4 inline with the current schema (working on it at this very minute).
a few more methods are being added.

 - Setting/Getting LR Power Level
 - Toggle RF
 - Node Info Received (NIF) event
 
 I am also setting the `ResultPayload` value to the `success` property if it exists at the `result`,
 as I sometimes set it to that, but others I don't - so want to make it consistent
 
 ```csharp
  internal CMDResult(JObject Res)
        {
            this.Success = Res.Value<bool>("success");

            if (Res.ContainsKey("result"))
            {
                bool? success = Res.SelectToken("result.success")?.Value<bool?>();
                if (success.HasValue)
                    SetPayload(success.Value);
            }

            if(Res.ContainsKey("zwaveErrorCode"))
                this.ErrorCode = Res.Value<string>("zwaveErrorCode");

            if (Res.ContainsKey("zwaveErrorMessage"))
                this.Message = Res.Value<string>("zwaveErrorMessage");
        }
 ```
 
Meaning if you don't use `SetPayload` it will use `result.success` if it exists.
`.success` is the transaction its self. -so I think this needs clearing  up.

Will tag you once those changes are pushed, so you can grab the latest v4 code

marcus-j-davies commented:
@spudwebb 

I am adding a new method to the `CMDResult` class in V4 - which removes the need to do manual casting (I find it annoying)

```csharp
public T ResultPayloadAs<T>()
{
     return ResultPayload is T value ? value : default!;
}
```


Usage
```csharp
Driver.Controller.Nodes.get(4).GetDefinedValueIDs().ContinueWith((R) =>{

    ValueID[] VIDs = R.Result.ResultPayloadAs<ValueID[]>()
    
    foreach(ValueID VID in VIDs){
         // Do something with VID (Value ID)
    }
})
```

Once again, I will ping you once V4 has been updated with all my changes & bumps to the schema

marcus-j-davies committed:
Bump

marcus-j-davies commented:
@spudwebb 

Do you want to now pull the latest V4 sources from  the V4 branch?

 - Driver version : `15.20.0`
 - PSI Server Version : `3.5.0` (schema 44)
 - Added method (CMDResult) `public T ResultPayloadAs<T>()` - Removes the need to cast the result yourself
 - Added method (Controller) `SetMaxLongRangePowerlevel`
 - Added method (Controller) `GetMaxLongRangePowerlevel`
 - Added method  (Controller)`ToggleRF`
 - Added event (ZWaveNode) `NodeInfoEvent`
 - Some small result path corrections
 
 Let me know if any issue.
 This commit will likely be the last before I drop support for older versions of the .NET SDK and the embedded source of the WS lib.
 
 I don't think this applies to you, but still:
 
 To build the PSI
 
  - Download a clean copy of the branch
   - With in the `PSI` folder, issue the following
     - `npm install`
     - `npm run build`
     - Rename `sever.exe` to `server.psi`

spudwebb commented:
Aren't you going to merge and release a v4 version before you switch to v5 and drop support for older versions of the .NET SDK ?

marcus-j-davies commented:
Merged - Will be releasing to nuget soon.

marcus-j-davies commented:
Published to NuGet!

It will be a lil while before the main branch has the V5 commits, so any issues with this release, do let me know, and I'll publish fixes as needed.

I would grab a copy of the main branch as it is now, so you can maintain backwards compatibility.
